### ISSUE TITLE
Processing one by one / avoiding slow searches of big mailbox.

### ISSUE THREAD

**Issue by divdot** [lefcha/imapfilter#116]
Hi.
I want to process only the latest messages, but `is_new/recent()` is not usable since I have other clients reading my mailbox, and `is_newer(1)` is way to slow since I have 1000+ messages in my inbox. (I want to run the filter every minute at least, to prefilter for my mobile)
I should reorganize my inbox, you may say, but I think there should be a smarter way.

I have hit upon the idea to get `inbox[uidnext-1]` and work backwards. This is fast and independent of mailbox size. I can keep track of the largest uid that was not moved and stop processing there.
Skipping uids that don't exist anymore is easy.

The snag with this plan, is that filters and processing only work with result lists.
The filtering is still doable, but the processing not so much.

As I am rather new to Lua, I can not immediately see a way to solve this.
I think the ideal solution would be if it was possible to get a  slice/ range from inbox.
I know there is no such thing in Lua, and I don't know if IMAP has support for such a concept..
It would have to be something like `result = inbox.get_range(lastprocessed : uidnext-i)`

If IMAP does not support slices/ ranges of uids, it might be simpler if it was possible to create an empty result list, and then add messages to it before filtering and moving.
If I was a Lua wizard I would probably have figured out how to do this by now by inspecting the code...

So.. Am I on the right track here? Is any of this doable "as is", or would you be willing to add the necessary method(s)?

**Comment by lefcha** [comment]
Hello,

First you should have a look at the IDLE extension, which is supported by imapfilter with the `enter_idle()` function, and there are also 2-3 options to fine-tune it.  Maybe it would fit you, as I think regardless of which client first sees the new messages, all clients connected and in IDLE mode also receive a notification about them.  Another benefit is that you don't need to poll every minute, as in IDLE mode you get informed when a message has arrived.

Other than that, and irregardless of the way you want to move forward, I wanted to let you know that you can actually create an empty results set, and then insert data (ie. the mailbox and message) into it.  There are some examples of exactly that in the [extend.lua](https://github.com/lefcha/imapfilter/blob/master/samples/extend.lua) file.  Have a look there, and if you need more help, please ask again here.

**Comment by divdot** [comment]
IDLE.. That's right, I should use that. Much better than polling. Thanks.
But the info returned, `n EXISTS` and `n RECENT` seems nearly useless from my point of view.
Probably why you did not bother to have `enter_idle()` return them either :)
So with regards to finding and filtering the new messages, it looks like IDLE makes no difference from polling.

However I realized yesterday that `send_query("UID n:uidnext")` or `n:*` will work, and return a table, at least with my ISP's server. (Courier-IMAP)
The debug log* shows the query being sent as `UID SEARCH ALL UID n1:n2`, while I suspect it should be just `UID n1:n2` or `UID SEARCH UID n1:n2`, but I have not compiled from source so I have only been able to test the last of those, and I have just browsed the RFC so I could be wrong.

Combined with status and keeping track of uidnext, and `match_` instead of `contains_`  that should enable me to solve this in an efficient way that works for both polling and idle...

PS *
Is the ssl key available somewhere to make wireshark able to decrypt the imapfilter traffic, or is there another way to see the raw traffic on the wire? I was not able to disable ssl/ starttls.

**Comment by lefcha** [comment]
Well, IDLE is more resource friendly to the server, as you don't poll continuously, but you get informed when there's a new message.  If new mail arrives all the time then it's not much of a difference I suppose, but if you get email once in a while, IDLE should be preferred.

Note, in case you didn't notice and it's useful, that you can get the next UID as the fourth value returned by `check_status()`.

Other than that, sending a custom query with the messages after a specific UID is one way to work around this, but in the past I have also recommended another way to users that couldn't make use of the `\Recent` or `\Seen` flags that messages have when they are new in a mailbox.  There's a special flag that messages can have and it's called `\Flagged`, and you could use this to mark messages that you have already filtered, as new messages will never have this flag set.

There's a method to help you set this flag in imapfilter `mark_flagged()` (there's also `unmark_flagged()` to unset it), and you can search for these messages with `is_unflagged()` (or `is_flagged()` if you want to do the reverse).  So my proposition is to use `is_unflagged()` either after `enter_idle()` returns or whenever you poll, and then as the messages you get back are those that for sure haven't been filtered, you process them anyway you like, and finally you mark them as flagged with `mark_flagged()`.  This final step makes all the difference, and helps you mark the messages as "done".

Regarding the encrypted traffic, whenever I want to play with some server and I have to use SSL/TLS, I just use something like `openssl s_client -host <your-encrypted-imap-server> -port 993 -crlf`, and with that I can send any IMAP command as I would do with `telnet` on an unencrypted connection.

**Comment by divdot** [comment]
I'm indeed using the uidnext value from check_status(). Without that I'd have had to use some totally different approach like the one you suggest using a flag or tag.
The trouble with that is that it still involves a serverside search of the entire inbox, something which in my case can take as much as 15 seconds.
The query for UIDs on the other hand, returns almost instantly.
(a meta search for contain_from, on those results adds another 15 second roundtrip, which is why I use match_from instead. Obviously the serverside DB could do with a few more indexes.)

Anyway I have now had a solution running for some days that seems stable and dependable, and works about like this:
```
read qrystart from UIDfile.
while idle
  get uidnext
  write UIDfile
  if qrystart and qrystart ~= uidnext
    send_query(qrystart:uidnext-1)
    qrystart=uidnext
    dofiltering
    idle = enter_idle
  end
end
```

The only question remaining to me is if the IMAP query mentioned earlier is correct, or if it just happens to work with my particular mailserver. 

Re. ssl testing, that looks like a useful trick. Thanks. Now what I wanted to do was explore what imapfilter was doing, not the server at the other end :-) Does not imapfilter write its session keys to disk somewhere?

**Comment by divdot** [comment]
PS.
When I loose connectivity while in idle I get this
`imapfilter: timeout period expired while waiting to read data`.
However it does not appear that the connection is resumed.
Should it do that, or is there some way for me to catch this and initiate a new idle when the connectivity is back. Probably just my unfamiliarity with Lua.. :-)

**Comment by divdot** [comment]
Did a quick check in python:
```
1006 UID 38368:38397
1006 NO Error in IMAP command received by server.

1007 UID SEARCH 38368:38397
* SEARCH
1007 OK SEARCH done.

1008 UID SEARCH UID 38368:38397
* SEARCH 38379 38384 38388 38389 38394 38395
1008 OK SEARCH done.

1009 UID SEARCH ALL UID 38368:38397
* SEARCH 38379 38384 38388 38389 38394 38395
1009 OK SEARCH done.
```
No idea whether it's more correct with the ALL or without, or if it matters at all..?
It's the `default initial key for ANDing`.. does that mean it's implicit if omitted?
Hard to find info in this..

**Comment by lefcha** [comment]
According to the IMAP [RFC 3501](http://tools.ietf.org/rfcmarkup?doc=3501), and assuming your last message was `123` and the next UID is `145` you can use either something like `123:145` or `123:*`. This means that you actually don't need the `UIDNEXT`, as the latter form doesn't require it, so only keeping the last processed message UID should be adequate.  Your query should work on all servers that are standards compliant.

Also, regarding including `ALL` some servers seem to require it and some don't so you can try both ways (I think I should add an option for that, as IIRC at least one server didn't like `ALL` when other search keys were specified).  According to the specs `ALL` or a `<sequence set>` should be defined if no other search keys are part of the search query (eg. `SEARCH ALL` or `SEARCH 123:*`), and if more search keys are defined, then it's correct to either have `ALL` in there or not, although it's really not needed.  So `SEARCH ALL FLAGGED` or `SEARCH FLAGGED` is basically the same query.

Now if you get a timeout with IDLE you should try lowering the `options.keepalive` option (see the imapfilter_config(5) man page), since the default is `29`, so try `19` and then `9`.

**Comment by divdot** [comment]
Hm.. I read about the 123:* form earlier. The issue I perceived with that is that `the contents of a range are independent of the order of the range endpoints.` which supposedly means that "(uidnext+100):*" will return one message, which I find a bit messy. On actually testing it now however I find that my ISPs server both look at the order and happily return empty results. Anyway I guess that the worst that would happen is that one would re-process one uid on startup (or add the code to avoid it), as on return from IDLE you would always get at least one "real" result, and thereby avoid the "magical" one.
After playing with this a bit it turns out the code is actually a few lines smaller, and avoiding a check_status call on every return from IDLE must be a good thing(tm).

Re. ALL. Thanks for the explanation. That makes sense. I should be fine then :)

Re. timeout. I actually made a quick bash script to time the interval between timeouts from my ISP and found it to be 5min and 5 sec. keepalive = 5 solved that nicely. What I tried to ask about however was if there is a way to handle timeouts resulting from actual network glitches? Say the wifi was down for an hour. Is there a way to catch that, and restart the connection?

**Comment by lefcha** [comment]
Currently imapfilter cannot recover from multiple network errors, such as those that would take place when the network is down, and instead it aborts with some error messages.  This means that the only way around this is an external script that would try to run imapfilter in a loop.

In theory it could be possible to implement this infinite (or limited by a number of tries) recover functionality, and I'll look into it when I have some time...

**Comment by lefcha** [comment]
Just added a new option called `options.persist`, which makes it possible to retry to recover a lost session indefinitely in case of temporary network errors.

**Comment by lefcha** [comment]
Closed by 0235a75be7685f64273353fdba401b1d1c6ca40e.
